spec/system/admin/dfc_product_import_spec.rb

```diff
@@ -74,6 +74,27 @@
     expect(product.image).to be_present
   end
 
+  it "shows oauth error message", vcr: true do
+    allow_any_instance_of(DfcRequest).to receive(:refresh_access_token!).and_raise(
+      Rack::OAuth2::Client::Error.new(
+        1, { error: "invalid_grant", error_description: "session not active" }
+      )
+    )
```

Comment:
Interesting, it's the call to Keycloak which raises the error. Nice to catch that but I'm wondering if we should take further action. I think that `refresh_access_token!` should delete the current refresh_token from the database. The OIDC settings view accounts for that already:

https://github.com/openfoodfoundation/openfoodnetwork/blob/a1df61c6125ea74e6f454ae082c2fb4db5b9baf9/app/views/admin/oidc_settings/index.html.haml#L21-L29

The error message could then have a link to the OIDC settings page. Raising a custom error in this case would make the code a bit easier to understand as well. Then you see what that OAuth2 error means.

Reply:
Thanks, I gave that a shot. I knew it would be a slight challenge adding the translation.. it turned out to be harder than I thought!

Sharing my learning: it would have been worth switching `config.action_controller.perform_caching` off 
so that I could try different things mid-test.